File: Sources/Views/Settings/Panes/CodeSpeakSettingsPane.swift
// MARK: - CodeSpeakSettingsPane
// Settings pane for the CodeSpeak spec-driven AI coding tool.
// macOS 14+, Swift 5.10

import SwiftUI

/// Settings pane for CodeSpeak.
///
/// Shows installation status, behaviour preferences, and API key info.
struct CodeSpeakSettingsPane: View {

    @Environment(\.csPreferences) private var csPreferences

    @State private var isInstalled: Bool = false
    @State private var hasAPIKey: Bool = false

    var body: some View {
        ScrollView(.vertical, showsIndicators: true) {
            VStack(alignment: .leading, spacing: DSSpacing.xl) {
                Text("CodeSpeak")
                    .font(DSFont.settingsTitle)
                    .foregroundStyle(DSColor.textPrimary)

                Divider().background(DSColor.borderDefault)

                statusSection
                behaviourSection
                displaySection
                notificationsSection
                apiKeySection

                if !isInstalled {
                    installSection
                }
            }
            .padding(DSSpacing.xl)
            .frame(maxWidth: .infinity, alignment: .topLeading)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .task {
            isInstalled = CLIAgentPathResolver.resolve("codespeak") != nil
            hasAPIKey = resolveHasAuth()
        }
    }

    // MARK: - Auth Resolution

    /// Checks both codespeak OAuth token and ANTHROPIC_API_KEY (Keychain + process env).
    /// CodeSpeak can authenticate via `codespeak login` (OAuth) OR via raw API key —
    /// we show "configured" if either is present.
    private func resolveHasAuth() -> Bool {
        if hasCodeSpeakToken() { return true }
        if let key = KeychainHelper.load(account: "ANTHROPIC_API_KEY"), !key.isEmpty { return true }
        if let key = ProcessInfo.processInfo.environment["ANTHROPIC_API_KEY"], !key.isEmpty { return true }
        return false
    }

    /// Returns true if `~/.codespeak/token.json` contains at least one access token.
    private func hasCodeSpeakToken() -> Bool {
        let url = URL(fileURLWithPath: NSHomeDirectory())
            .appending(path: ".codespeak/token.json")
        guard let data = try? Data(contentsOf: url),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let tokens = json["tokens"] as? [[String: Any]] else { return false }
        return tokens.contains { ($0["access_token"] as? String)?.isEmpty == false }
    }

    // MARK: - Status

    private var statusSection: some View {
        VStack(alignment: .leading, spacing: DSSpacing.sm) {
            SettingsSectionHeader(title: "Status")

            VStack(spacing: 0) {
                statusRow(
                    ok: isInstalled,
                    okText: "codespeak installed",
                    failText: "codespeak not found"
                )
                Divider()
                    .background(DSColor.borderSubtle)
                    .padding(.horizontal, DSSpacing.md)
                statusRow(
                    ok: hasAPIKey,
                    okText: "Authenticated",
                    failText: "Not authenticated (run codespeak login)",
                    failColor: DSColor.gitModified
                )
            }
            .settingsCard()
        }
    }

    // MARK: - Behaviour

    private var behaviourSection: some View {
        VStack(alignment: .leading, spacing: DSSpacing.sm) {
            SettingsSectionHeader(title: "Behaviour")

            VStack(alignment: .leading, spacing: 0) {
                preferenceRow(
                    title: "Auto-build on save",
                    description: "Run `codespeak build` automatically after saving a spec file",
                    isOn: Binding(
                        get: { csPreferences.autoBuildOnSave },
                        set: { csPreferences.autoBuildOnSave = $0 }
                    )
                )

                Divider().padding(.leading, DSSpacing.md)

                preferenceRow(
                    title: "Build on project open",
                    description: "Run `codespeak build` when switching to a CodeSpeak project",
                    isOn: Binding(
                        get: { csPreferences.buildOnProjectOpen },
                        set: { csPreferences.buildOnProjectOpen = $0 }
                    )
                )

                Divider().padding(.leading, DSSpacing.md)

                preferenceRow(
                    title: "Auto-open build panel",
                    description: "Open the build panel automatically when a command starts",
                    isOn: Binding(
                        get: { csPreferences.autoOpenBuildPanel },
                        set: { csPreferences.autoOpenBuildPanel = $0 }
                    )
                )

                Divider().padding(.leading, DSSpacing.md)

                defaultCommandRow
            }
            .settingsCard()
        }
    }

    // MARK: - Default Command Row

    private var defaultCommandRow: some View {
        HStack(spacing: DSSpacing.sm) {
            VStack(alignment: .leading, spacing: DSSpacing.xxs) {
                Text("Default command")
                    .font(DSFont.buttonLabel)
                    .foregroundStyle(DSColor.textPrimary)
                Text("Command executed by the ▶ button")
                    .font(DSFont.sidebarItemSmall)
                    .foregroundStyle(DSColor.textSecondary)
            }

            Spacer()

            Picker("", selection: Binding(
                get: { csPreferences.defaultCommand },
                set: { csPreferences.defaultCommand = $0 }
            )) {
                // .task and .change require text input — exclude from silent defaults
                ForEach(CodeSpeakCommand.allCases.filter { !$0.requiresInput }) { cmd in
                    Text(cmd.displayName).tag(cmd)
                }
            }
            .pickerStyle(.menu)
            .frame(width: 90)
            .labelsHidden()
        }
        .padding(.horizontal, DSSpacing.md)
        .padding(.vertical, DSSpacing.sm)
    }

    // MARK: - Display

    private var displaySection: some View {
        VStack(alignment: .leading, spacing: DSSpacing.sm) {
            SettingsSectionHeader(title: "Display")

            VStack(alignment: .leading, spacing: 0) {
                preferenceRow(
                    title: "Show failing specs only",
                    description: "Filter the spec list to show only specs with failing status",
                    isOn: Binding(
                        get: { csPreferences.showFailingOnly },
                        set: { csPreferences.showFailingOnly = $0 }
                    )
                )
            }
            .settingsCard()
        }
    }

    // MARK: - Notifications

    private var notificationsSection: some View {
        VStack(alignment: .leading, spacing: DSSpacing.sm) {
            SettingsSectionHeader(title: "Notifications")

            VStack(alignment: .leading, spacing: 0) {
                preferenceRow(
                    title: "Notify on complete",
                    description: "Send a macOS notification when a command finishes",
                    isOn: Binding(
                        get: { csPreferences.notifyOnComplete },
                        set: { csPreferences.notifyOnComplete = $0 }
                    )
                )
            }
            .settingsCard()
        }
    }

    // MARK: - API Key

    private var apiKeySection: some View {
        VStack(alignment: .leading, spacing: DSSpacing.sm) {
            SettingsSectionHeader(title: "API Key")
            Text("Authenticate via `codespeak login` (OAuth) or set ANTHROPIC_API_KEY in Settings → Claude.")
                .font(DSFont.buttonLabel)
                .foregroundStyle(DSColor.textSecondary)
                .fixedSize(horizontal: false, vertical: true)
                .padding(DSSpacing.md)
                .frame(maxWidth: .infinity, alignment: .leading)
                .settingsCard()
        }
    }

    // MARK: - Installation

    private var installSection: some View {
        VStack(alignment: .leading, spacing: DSSpacing.sm) {
            SettingsSectionHeader(title: "Installation")

            Text("""
                1. Install uv: curl -LsSf https://astral.sh/uv/install.sh | sh
                2. uv tool install codespeak-cli
                3. codespeak init  (in your project)
                """)
                .font(DSFont.terminal(size: 12))
                .foregroundStyle(DSColor.textPrimary)
                .padding(DSSpacing.md)
                .frame(maxWidth: .infinity, alignment: .leading)
                .settingsCard()
        }
    }

    // MARK: - Helpers

    private func statusRow(
        ok: Bool,
        okText: String,
        failText: String,
        failColor: Color = DSColor.gitDeleted
    ) -> some View {
        HStack(spacing: DSSpacing.xs) {
            Circle()
                .fill(ok ? DSColor.gitAdded : failColor)
                .frame(width: DSLayout.statusDotSize, height: DSLayout.statusDotSize)
            Text(ok ? okText : failText)
                .font(DSFont.buttonLabel)
                .foregroundStyle(DSColor.textPrimary)
        }
        .padding(.horizontal, DSSpacing.md)
        .padding(.vertical, DSSpacing.sm)
    }

    private func preferenceRow(
        title: String,
        description: String,
        isOn: Binding<Bool>
    ) -> some View {
        HStack(spacing: DSSpacing.sm) {
            VStack(alignment: .leading, spacing: DSSpacing.xxs) {
                Text(title)
                    .font(DSFont.buttonLabel)
                    .foregroundStyle(DSColor.textPrimary)
                Text(description)
                    .font(DSFont.sidebarItemSmall)
                    .foregroundStyle(DSColor.textSecondary)
                    .fixedSize(horizontal: false, vertical: true)
            }

            Spacer()

            Toggle("", isOn: isOn)
                .toggleStyle(.switch)
                .labelsHidden()
                .controlSize(.small)
        }
        .padding(.horizontal, DSSpacing.md)
        .padding(.vertical, DSSpacing.sm)
    }
}

File: desktopApp/src/main/kotlin/studio/vibe/desktop/ui/settings/CodeSpeakSettingsPane.kt
package studio.vibe.desktop.ui.settings

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ExposedDropdownMenuBox
import androidx.compose.material3.ExposedDropdownMenuDefaults
import androidx.compose.material3.ExposedDropdownMenuAnchorType
import androidx.compose.material3.Text
import androidx.compose.material3.TextField
import androidx.compose.material3.TextFieldDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import studio.vibe.desktop.DesktopServiceContainer
import studio.vibe.desktop.ui.theme.DSColor
import studio.vibe.desktop.ui.theme.DSFont
import studio.vibe.desktop.ui.theme.DSRadius
import studio.vibe.desktop.ui.theme.DSSpacing
import studio.vibe.shared.model.CodeSpeakCommand
import java.io.File

/**
 * Settings pane for CodeSpeak spec-driven AI coding tool.
 *
 * Mirrors Swift CodeSpeakSettingsPane.swift.
 *
 * Shows:
 * - Status (installed / authenticated)
 * - Behaviour preferences (auto-build, build-on-open, auto-open panel, default command)
 * - Display preferences (show failing only)
 * - Notifications (notify on complete)
 * - API key info
 * - Installation instructions (when not installed)
 */
@Composable
fun CodeSpeakSettingsPane(
    container: DesktopServiceContainer,
    modifier: Modifier = Modifier,
) {
    val prefs = container.codeSpeakPreferences

    // Installation status
    var isInstalled by remember { mutableStateOf(false) }
    var hasAPIKey by remember { mutableStateOf(false) }

    // Behaviour prefs
    var autoBuildOnSave by remember { mutableStateOf(prefs.autoBuildOnSave) }
    var buildOnProjectOpen by remember { mutableStateOf(prefs.buildOnProjectOpen) }
    var autoOpenBuildPanel by remember { mutableStateOf(prefs.autoOpenBuildPanel) }
    var defaultCommand by remember { mutableStateOf(prefs.defaultCommand) }

    // Display prefs
    var showFailingOnly by remember { mutableStateOf(prefs.showFailingOnly) }

    // Notification prefs
    var notifyOnComplete by remember { mutableStateOf(prefs.notifyOnComplete) }

    LaunchedEffect(Unit) {
        // Check if codespeak binary is on PATH
        isInstalled = runCatching {
            val result = ProcessBuilder("which", "codespeak")
                .start()
                .waitFor()
            result == 0
        }.getOrDefault(false)

        // Check for codespeak token file or ANTHROPIC_API_KEY
        hasAPIKey = runCatching {
            val tokenFile = File("${System.getProperty("user.home")}/.codespeak/token.json")
            if (tokenFile.exists()) {
                true
            } else {
                System.getenv("ANTHROPIC_API_KEY")?.isNotEmpty() == true
            }
        }.getOrDefault(false)
    }

    Column(
        modifier = modifier
            .verticalScroll(rememberScrollState())
            .padding(DSSpacing.xl),
        verticalArrangement = Arrangement.spacedBy(DSSpacing.xl),
    ) {
        SettingsPaneTitle(title = "CodeSpeak")

        // -- Status ----------------------------------------------------------

        Column(
            modifier = Modifier.fillMaxWidth(),
            verticalArrangement = Arrangement.spacedBy(DSSpacing.sm),
        ) {
            SettingsSectionHeader(title = "Status")

            SettingsCard {
                Column {
                    StatusRow(
                        ok = isInstalled,
                        okText = "codespeak installed",
                        failText = "codespeak not found",
                    )
                    SettingsListDivider()
                    StatusRow(
                        ok = hasAPIKey,
                        okText = "Authenticated",
                        failText = "Not authenticated — run codespeak login",
                        failColor = DSColor.gitModified,
                    )
                }
            }
        }

        // -- Behaviour -------------------------------------------------------

        Column(
            modifier = Modifier.fillMaxWidth(),
            verticalArrangement = Arrangement.spacedBy(DSSpacing.sm),
        ) {
            SettingsSectionHeader(title = "Behaviour")

            SettingsCard {
                Column {
                    SettingsToggleRow(
                        title = "Auto-build on save",
                        description = "Run `codespeak build` automatically after saving a spec file",
                        checked = autoBuildOnSave,
                        onCheckedChange = { value ->
                            autoBuildOnSave = value
                            prefs.autoBuildOnSave = value
                        },
                    )

                    SettingsListDivider()

                    SettingsToggleRow(
                        title = "Build on project open",
                        description = "Run `codespeak build` when switching to a CodeSpeak project",
                        checked = buildOnProjectOpen,
                        onCheckedChange = { value ->
                            buildOnProjectOpen = value
                            prefs.buildOnProjectOpen = value
                        },
                    )

                    SettingsListDivider()

                    SettingsToggleRow(
                        title = "Auto-open build panel",
                        description = "Open the build panel automatically when a command starts",
                        checked = autoOpenBuildPanel,
                        onCheckedChange = { value ->
                            autoOpenBuildPanel = value
                            prefs.autoOpenBuildPanel = value
                        },
                    )

                    SettingsListDivider()

                    // Default command picker
                    DefaultCommandRow(
                        selectedCommand = defaultCommand,
                        onCommandSelected = { cmd ->
                            defaultCommand = cmd
                            prefs.defaultCommand = cmd
                        },
                    )
                }
            }
        }

        // -- Display ---------------------------------------------------------

        Column(
            modifier = Modifier.fillMaxWidth(),
            verticalArrangement = Arrangement.spacedBy(DSSpacing.sm),
        ) {
            SettingsSectionHeader(title = "Display")

            SettingsCard {
                SettingsToggleRow(
                    title = "Show failing specs only",
                    description = "Filter the spec list to show only specs with failing status",
                    checked = showFailingOnly,
                    onCheckedChange = { value ->
                        showFailingOnly = value
                        prefs.showFailingOnly = value
                    },
                )
            }
        }

        // -- Notifications ----------------------------------------------------

        Column(
            modifier = Modifier.fillMaxWidth(),
            verticalArrangement = Arrangement.spacedBy(DSSpacing.sm),
        ) {
            SettingsSectionHeader(title = "Notifications")

            SettingsCard {
                SettingsToggleRow(
                    title = "Notify on complete",
                    description = "Show a system notification when a command finishes",
                    checked = notifyOnComplete,
                    onCheckedChange = { value ->
                        notifyOnComplete = value
                        prefs.notifyOnComplete = value
                    },
                )
            }
        }

        // -- API Key ----------------------------------------------------------

        Column(
            modifier = Modifier.fillMaxWidth(),
            verticalArrangement = Arrangement.spacedBy(DSSpacing.sm),
        ) {
            SettingsSectionHeader(title = "API Key")

            SettingsCard {
                Text(
                    text = "Authenticate via `codespeak login` (OAuth) or set ANTHROPIC_API_KEY in Settings → Claude.",
                    style = DSFont.buttonLabel,
                    color = DSColor.textSecondary,
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(horizontal = DSSpacing.md, vertical = DSSpacing.sm),
                )
            }
        }

        // -- Installation (shown only when not installed) ---------------------

        if (!isInstalled) {
            Column(
                modifier = Modifier.fillMaxWidth(),
                verticalArrangement = Arrangement.spacedBy(DSSpacing.sm),
            ) {
                SettingsSectionHeader(title = "Installation")

                SettingsCard {
                    Text(
                        text = "1. Install uv: curl -LsSf https://astral.sh/uv/install.sh | sh\n" +
                            "2. uv tool install codespeak-cli\n" +
                            "3. codespeak init  (in your project)",
                        style = DSFont.monoSmall,
                        color = DSColor.textPrimary,
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(horizontal = DSSpacing.md, vertical = DSSpacing.sm),
                    )
                }
            }
        }

        Spacer(Modifier.height(DSSpacing.xl))
    }
}

// -- Private helper composables -------------------------------------------------

@Composable
private fun StatusRow(
    ok: Boolean,
    okText: String,
    failText: String,
    failColor: Color = DSColor.gitDeleted,
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = DSSpacing.md, vertical = DSSpacing.sm),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Box(
            modifier = Modifier
                .size(8.dp)
                .background(
                    color = if (ok) DSColor.indicatorRunning else failColor,
                    shape = CircleShape,
                ),
        )
        Spacer(Modifier.width(DSSpacing.sm))
        Text(
            text = if (ok) okText else failText,
            style = DSFont.buttonLabel,
            color = DSColor.textPrimary,
        )
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun DefaultCommandRow(
    selectedCommand: CodeSpeakCommand,
    onCommandSelected: (CodeSpeakCommand) -> Unit,
) {
    val validCommands = CodeSpeakCommand.entries.filter { !it.requiresInput }
    var expanded by remember { mutableStateOf(false) }

    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = DSSpacing.md, vertical = DSSpacing.sm),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Column(modifier = Modifier.weight(1f)) {
            Text(
                text = "Default command",
                style = DSFont.buttonLabel,
                color = DSColor.textPrimary,
            )
            Text(
                text = "Command executed by the run button",
                style = DSFont.sidebarItemSmall,
                color = DSColor.textSecondary,
            )
        }

        Spacer(Modifier.width(DSSpacing.sm))

        ExposedDropdownMenuBox(
            expanded = expanded,
            onExpandedChange = { expanded = it },
        ) {
            TextField(
                value = selectedCommand.displayName,
                onValueChange = {},
                readOnly = true,
                trailingIcon = { ExposedDropdownMenuDefaults.TrailingIcon(expanded = expanded) },
                modifier = Modifier
                    .menuAnchor(ExposedDropdownMenuAnchorType.PrimaryNotEditable)
                    .width(100.dp),
                textStyle = DSFont.bodySmall.copy(color = DSColor.textPrimary),
                colors = TextFieldDefaults.colors(
                    focusedContainerColor = DSColor.surfaceOverlay,
                    unfocusedContainerColor = DSColor.surfaceOverlay,
                    focusedIndicatorColor = Color.Transparent,
                    unfocusedIndicatorColor = Color.Transparent,
                ),
            )

            ExposedDropdownMenu(
                expanded = expanded,
                onDismissRequest = { expanded = false },
            ) {
                validCommands.forEach { cmd ->
                    DropdownMenuItem(
                        text = {
                            Text(
                                text = cmd.displayName,
                                style = DSFont.bodySmall,
                                color = DSColor.textPrimary,
                            )
                        },
                        onClick = {
                            onCommandSelected(cmd)
                            expanded = false
                        },
                    )
                }
            }
        }
    }
}
